collect_app/src/main/java/org/odk/collect/android/logic/actions/setlocation/CollectSetLocationAction.java

```diff
@@ -0,0 +1,100 @@
+
+import android.location.Location;
+
+import com.google.android.gms.location.LocationListener;
+
+import org.javarosa.core.model.actions.setlocation.SetLocationAction;
+import org.javarosa.core.model.instance.TreeReference;
+import org.odk.collect.android.application.Collect;
+import org.odk.collect.android.location.client.GoogleLocationClient;
+import org.odk.collect.android.location.client.MaxAccuracyWithinTimeoutLocationClient;
+import org.odk.collect.android.location.client.LocationClients;
+import org.odk.collect.android.preferences.GeneralSharedPreferences;
+import org.odk.collect.android.utilities.GeoUtils;
+
+import timber.log.Timber;
+
+import static org.odk.collect.android.preferences.GeneralKeys.KEY_BACKGROUND_LOCATION;
+
+/**
+ * An Android-specific implementation of {@link SetLocationAction}.
+ *
+ * When the action is triggered, the first location available is saved and the highest-accuracy
+ * reading within {@link #SECONDS_TO_CONSIDER_UPDATES} seconds will replace it. If no location is
+ * available within {@link #SECONDS_TO_CONSIDER_UPDATES} seconds, no location is written.
+ *
+ * It is possible for the same action to be triggered multiple times within
+ * {@link #SECONDS_TO_CONSIDER_UPDATES} seconds if the action is triggered by a value changed
+ * event, for example. In that case, the highest accuracy reading resets each time the action is
+ * triggered.
+ *
+ * In a repeat, the target node for subsequent times the action is triggered could be different than
+ * the target node the first time it was triggered. In that case, only the current target node is
+ * updated.
+ */
+public class CollectSetLocationAction extends SetLocationAction implements LocationListener {
+    private static final int SECONDS_TO_CONSIDER_UPDATES = 20;
+
+    private MaxAccuracyWithinTimeoutLocationClient maxAccuracyLocationClient;
+
+    public CollectSetLocationAction() {
+        // For serialization
+    }
+
+    // Needed to set the action name.
+    CollectSetLocationAction(TreeReference targetReference) {
+        super(targetReference);
+    }
+
+    @Override
+    public void requestLocationUpdates() {
+        // Do initialization on first location request so the client doesn't need to be serialized
+        if (maxAccuracyLocationClient == null) {
+            maxAccuracyLocationClient = new MaxAccuracyWithinTimeoutLocationClient(new GoogleLocationClient(Collect.getInstance().getApplicationContext()), this);
+        }
+
+        // Only start acquiring location if the Collect preference allows it and Google Play
+        // Services are available. If it's not allowed, leave the target field blank.
+        if (GeneralSharedPreferences.getInstance().getBoolean(KEY_BACKGROUND_LOCATION, true)
+            && LocationClients.areGooglePlayServicesAvailable(Collect.getInstance().getApplicationContext())) {
+            maxAccuracyLocationClient.requestLocationUpdates(SECONDS_TO_CONSIDER_UPDATES);
```

Comment:
I really like that the detail of these updates is now all hidden inside the client!

Reply:
💃 Thanks for talking through that with me!